netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -797,101 +901,227 @@ _net_ebpf_extension_connection_context_initialize(
 
-    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Flink;
-    while (list_entry != &_net_ebpf_ext_connect_context_list) {
-        net_ebpf_extension_connection_context_t* entry =
-            CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
-        if (memcmp(
-                &local_connection_context, entry, EBPF_OFFSET_OF(net_ebpf_extension_connection_context_t, timestamp)) ==
-            0) {
-            // Found matching entry. Remove it from the list and return.
-            RemoveEntryList(&entry->list_entry);
-            _net_ebpf_ext_connect_context_count--;
-            connection_context = entry;
-            break;
-        }
-        list_entry = list_entry->Flink;
-    }
-
-    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_blocked_contexts.lock);
+    entry_found = _net_ebpf_ext_find_and_remove_connection_context_locked(&local_connection_context);
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_blocked_contexts.lock, old_irql);
 
-    NET_EBPF_EXT_RETURN_POINTER(net_ebpf_extension_connection_context_t*, connection_context);
+    return entry_found;
 }
 
-_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_lock) static void _net_ebpf_ext_purge_lru_contexts_under_lock(
-    BOOLEAN delete_all)
+_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_blocked_contexts
+                                   .lock) static void _net_ebpf_ext_purge_blocked_connect_contexts(bool delete_all)
 {
     uint64_t expiry_time = CONVERT_100NS_UNITS_TO_MS(KeQueryInterruptTime()) - EXPIRY_TIME;
 
-    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Blink;
-    while (list_entry != &_net_ebpf_ext_connect_context_list) {
+    // Free entries from the LRU list. These entries should also be removed from the table.
+    LIST_ENTRY* list_entry = _net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_lru_list.Blink;
+    while (list_entry != &_net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_lru_list) {
         net_ebpf_extension_connection_context_t* entry =
             CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
+        // Move pointer to next entry prior to removing the entry.
+        list_entry = list_entry->Blink;
+
         if (!delete_all && entry->timestamp > expiry_time) {
             break;
         }
-#pragma warning(suppress : 6001) /* list entry is non-null */
-        list_entry = list_entry->Blink;
-        RemoveEntryList(&entry->list_entry);
-
-        _net_ebpf_ext_connect_context_count--;
 
+#pragma warning(suppress : 6001) /* entry and list entry are non-null */
+        RemoveEntryList(&entry->list_entry);
+        BOOLEAN result =
+            RtlDeleteElementGenericTableAvl(&_net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_table, entry);
```

Comment:
do we also need to call `ExFreePool` for the deleted entries after deleting them from hash table?

Reply:
We should not need to - the documentation mentions that the free routine is called. 

https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddk/nf-ntddk-rtldeleteelementgenerictableavl